### PR TITLE
Normalization for Snowflake destination: added support for key pair authentication

### PR DESCRIPTION
**Pull request by sashaNeshcheret**
## What
In scope of [PR](https://github.com/airbytehq/airbyte/pull/14388) we want to add support for key pair authentication for snowflake destination. But before publishing new version for snowflake we need to update normalization and release airbyte version with updated normalization( normalization should be aware about new auth type). 

## How
After bumping normalization, merging  this PR and releasing airbyte i can proceed with [PR](https://github.com/airbytehq/airbyte/pull/14388)

## Recommended reading order
1.  `y.python`
2. `x.java`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by grubberr** [comment]
@sashaNeshcheret what about changelog ?

**Comment by sashaNeshcheret** [comment]
> @sashaNeshcheret what about changelog ?

Thanks for comment, changelog was updated.

**Review by grubberr — APPROVED** [review]

File: airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py [hunk 1]
```diff
@@ -219,6 +219,13 @@ def transform_snowflake(config: Dict[str, Any]):
             dbt_config["oauth_client_id"] = credentials["client_id"]
             dbt_config["oauth_client_secret"] = credentials["client_secret"]
             dbt_config["token"] = credentials["refresh_token"]
+        elif credentials.get("private_key"):
+            f = open("private_key_path.txt", "w")
+            f.write(credentials["private_key"])
+            f.close()
+            dbt_config["private_key_path"] = "private_key_path.txt"
+            if credentials.get("passphrase"):
+                dbt_config["private_key_passphrase"] = credentials["passphrase"]
```
Comment (evantahler) [inline airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py:228]: Noting that the authentication options that will be added to the Snowflake Destination will be called "private_key" and "passphrase".  Does this match other destinations?

**Review by evantahler — COMMENTED** [review]
I'm probably not a great person to review this, as I don't have the knowledge of what other DBs do to match the varaibles/style.  I'll remove myself from the list of reviewers

File: airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py (hunk 1 above)
Comment (sashaNeshcheret) [reply airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py:228]: "private_key" - is commonly used in other connectors,  "passphrase" - is just snowflake terminology, will rename it to password

**Comment by subodh1810** [comment]
@edgao can you review the PR, I think you are the right person since you have more knowledge in normalization

File: airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py [hunk 2]
```diff
@@ -219,6 +219,13 @@ def transform_snowflake(config: Dict[str, Any]):
             dbt_config["oauth_client_id"] = credentials["client_id"]
             dbt_config["oauth_client_secret"] = credentials["client_secret"]
             dbt_config["token"] = credentials["refresh_token"]
+        elif credentials.get("private_key"):
+            f = open("private_key_path.txt", "w")
```
Comment (edgao) [inline airbyte-integrations/bases/base-normalization/normalization/transform_config/transform.py:223]: nitpick: prefer `with open()... as f` over `open()... close()`; this is basically equivalent to Java's `try (OutputStream s = ...)`:
```python
with open("private_key_path.txt", "w") as f:
  f.write(credentials["private_key"])
```
(then you don't need to `f.close()` at all)

**Review by edgao — COMMENTED** [review]
running destination-snowflake tests. Can you run a manual test using a private key? I don't think normalization integration tests currently cover this (and I don't think they should).

I do think there should be a unit test for this though, similar to https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-normalization/unit_tests/test_transform_config.py#L304

**Comment by edgao** [comment]
also running bigquery+postgres per policy

**Comment by sashaNeshcheret** [comment]
Unit test was added. @edgao please take a look.

**Review by edgao — COMMENTED** [review]
nice! It looks like the unit test is failing though :( 

~also, bumping my question about running a manual test~ talked offline, this will happen as part of https://github.com/airbytehq/airbyte/pull/14388

**Review by edgao — APPROVED** [review]

**Pull request merged**